### PR TITLE
Sigma message construction and parsing for CASE

### LINKED ISSUE #1802: [Secure Channel] Operational Credentials
We need an API to utilize operational credentials. Specifically, we need to be able to:

1. Sign arbitrary blobs of data using the private key backing a device's op cred.
2. Export the operational credential (so as to be sent in a message).
3. Verify a given operational credential chains up to a root.

### LINKED ISSUE #4469: Implement Sigma message construction and parsing for CASE state machine
#### Problem
CASE state machine is barebones, and is missing Sigma message construction and parsing code.
 
 #### Proposed Solution
Implement Sigma message constructors and parser and use in CASE state machine.

### PR DESCRIPTION
**Pull request by jpk233**
#### Problem
CASE session functionality, along with sigma message protocol, is not implemented. Additionally, an operational certificate class is lacking for API usage.

 #### Summary of Changes
 * CASE session support added, currently only for single device certificates.
 * Sigma message construction and parsing implemented.
 * CHIP Operational cert data types added 

 Fixes #4469
 Fixes #1802

### CODE REVIEW

**Comment by woody-apple** [comment]
@pan-apple ?

**Comment by yufengwangca** [comment]
Currently, we are still using test shared key to test secure channel for TE1. I hope we can switch to CASE ASAP.

**Review by woody-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/credentials/CHIPOperationalCredentials.cpp
Comment (andy31415) [inline src/credentials/CHIPOperationalCredentials.cpp:58]: should we clear/detect if mOpCreds is already non-null? same for above init. This is to not allow Init to be called twice.

It may be easier to follow memory ownership if the 'array with/without ownership' would be a separate object type.

File: src/credentials/CHIPOperationalCredentials.h
Comment (andy31415) [inline src/credentials/CHIPOperationalCredentials.h:231]: std::map uses heap. Can we use some alternative?

**Review by bzbarsky-apple — COMMENTED** [review]
Unfortunately, this implemented the old (broken/underdefined in various ways) format of the CASE messages from the spec.  Now that https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/2010 has merged we should switch to the format defined there.

I don't know whether it's better to try merging this and then do a followup to rewrite all the parsing/serialization code, or just update this to align with the spec. @jpk233 thoughts?

File: src/credentials/CHIPOperationalCredentials.cpp
Comment (jpk233) [reply src/credentials/CHIPOperationalCredentials.cpp]: Thanks, updated.

File: src/credentials/CHIPOperationalCredentials.h
Comment (jpk233) [reply src/credentials/CHIPOperationalCredentials.h]: Agreed. Thanks, updated.

File: src/transport/CASESession.cpp
Comment (pan-apple) [inline src/transport/CASESession.cpp:45]: Are these hardcoded values temporary? Or, is it something that's defined in CHIP specifications?

File: src/transport/CASESession.cpp
Comment (pan-apple) [inline src/transport/CASESession.cpp:271]: Remove the commented out code?

File: src/transport/CASESession.cpp
Comment (pan-apple) [inline src/transport/CASESession.cpp:277]: Remove the commented out code?

File: src/transport/CASESession.cpp
Comment (pan-apple) [inline src/transport/CASESession.cpp:1046]: Add a TODO comment to remove it.

**Review by pan-apple — COMMENTED** [review]
Code structure looks good to me. As per @bzbarsky-apple comment, we might want to update it to match the spec.

File: src/credentials/CHIPOperationalCredentials.h
Comment (andy31415) [inline src/credentials/CHIPOperationalCredentials.h:61]: nit: delete `@class` delcaration as it is redundant

File: src/credentials/CHIPOperationalCredentials.h
Comment (andy31415) [inline src/credentials/CHIPOperationalCredentials.h:48]: maybe naming of `Pair` would be better than `Map`

**Review by andy31415 — COMMENTED** [review]
@jpk233 - let me know what you want to do about the comments from Boris and catching up with the updated spec.

My comments are nitpicks and I would not block the PR, but not yet approving so that it does not get merged before we get some answer. Let me know and if you want to merge this, I will approve and merge.

**Comment by jpk233** [comment]
@pan-apple Thanks, updated as per your comments.

@andy31415 I think we should move forward with this PR as it currently is, and address the spec update in a new PR.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**
